File: app/modules/pocket-module/ios/PocketModule.swift
import ExpoModulesCore
import PocketCore

public class PocketModule: Module {
  private let walletCore: CoreWalletCore = {
    guard let core = CoreNewWalletCore() else {
      preconditionFailure("Failed to initialize CoreWalletCore")
    }
    return core
  }()

  private let secureKeyStore = SecureKeyStore()

  public func definition() -> ModuleDefinition {
    Name("PocketCore")

    AsyncFunction("initWallet") { (dataDir: String, password: String, masterKeyB64: String, kdfSaltB64: String) in
      _ = try self.walletCore.init_(dataDir, password: password, masterKeyB64: masterKeyB64, kdfSaltB64: kdfSaltB64)
    }

    AsyncFunction("initWalletSecure") { (dataDir: String, password: String) in
      let masterKeyB64 = try self.secureKeyStore.getOrCreateMasterKeyBase64()
      let kdfSaltB64 = try self.secureKeyStore.getOrCreateKdfSaltBase64()
      _ = try self.walletCore.init_(dataDir, password: password, masterKeyB64: masterKeyB64, kdfSaltB64: kdfSaltB64)
    }

    AsyncFunction("closeWallet") {
      _ = try self.walletCore.close()
    }

    AsyncFunction("createEthereumWallet") { (name: String) throws -> String in
      var nsError: NSError?
      let value = self.walletCore.createEthereumWallet(name, error: &nsError)
      if let error = nsError { throw error }
      return value
    }

    AsyncFunction("openOrCreateWallet") { (name: String) throws -> String in
      var nsError: NSError?
      let value = self.walletCore.openOrCreateWallet(name, error: &nsError)
      if let error = nsError { throw error }
      return value
    }

    AsyncFunction("getBalance") { (network: String) throws -> String in
      var nsError: NSError?
      let value = self.walletCore.getBalance(network, error: &nsError)
      if let error = nsError { throw error }
      return value
    }

    AsyncFunction("getAccountSummary") { (network: String) throws -> String in
      var nsError: NSError?
      let value = self.walletCore.getAccountSummary(network, error: &nsError)
      if let error = nsError { throw error }
      return value
    }

    AsyncFunction("getAccountSnapshot") { (network: String) throws -> String in
      var nsError: NSError?
      let value = self.walletCore.getAccountSnapshot(network, error: &nsError)
      if let error = nsError { throw error }
      return value
    }

    AsyncFunction("createSmartContractAccount") { (network: String) throws -> String in
      var nsError: NSError?
      let value = self.walletCore.createSmartContractAccount(network, error: &nsError)
      if let error = nsError { throw error }
      return value
    }

    AsyncFunction("getSmartContractAccount") { (network: String) throws -> String in
      var nsError: NSError?
      let value = self.walletCore.getSmartContractAccount(network, error: &nsError)
      if let error = nsError { throw error }
      return value
    }

    AsyncFunction("listAccounts") { () throws -> String in
      var nsError: NSError?
      let value = self.walletCore.listAccounts(&nsError)
      if let error = nsError { throw error }
      return value
    }

    AsyncFunction("sendUsdc") { (network: String, destination: String, amount: String, note: String, providerID: String) throws -> String in
      var nsError: NSError?
      let value = self.walletCore.sendUsdc(network, destination: destination, amount: amount, note: note, providerID: providerID, error: &nsError)
      if let error = nsError { throw error }
      return value
    }

    AsyncFunction("sendToken") { (network: String, tokenIdentifier: String, destination: String, amount: String, note: String, providerID: String) throws -> String in
      var nsError: NSError?
      let value = self.walletCore.sendToken(network, tokenIdentifier: tokenIdentifier, destination: destination, amount: amount, note: note, providerID: providerID, error: &nsError)
      if let error = nsError { throw error }
      return value
    }

   AsyncFunction("getUsdcTransactions") { (network: String, limit: Int, offset: Int) throws -> String in
      var nsError: NSError?
      let value = self.walletCore.listUsdcTransactions(network, limit: limit, offset: offset, error: &nsError)
      if let error = nsError { throw error }
      return value
    }

    AsyncFunction("getTokenTransactions") { (network: String, tokenIdentifier: String, limit: Int, offset: Int) throws -> String in
      var nsError: NSError?
      let value = self.walletCore.listTokenTransactions(network, tokenIdentifier: tokenIdentifier, limit: limit, offset: offset, error: &nsError)
      if let error = nsError { throw error }
      return value
    }

    AsyncFunction("listAllTransactions") { (network: String, limit: Int, offset: Int) throws -> String in
      var nsError: NSError?
      let value = self.walletCore.listAllTransactions(network, limit: limit, offset: offset, error: &nsError)
      if let error = nsError { throw error }
      return value
    }

    AsyncFunction("exportBackup") { (passphrase: String) throws -> String in
      var nsError: NSError?
      let value = self.walletCore.exportWalletBackup(passphrase, error: &nsError)
      if let error = nsError { throw error }
      return value
    }

    AsyncFunction("importBackup") { (payload: String, passphrase: String) throws -> String in
      var nsError: NSError?
      let value = self.walletCore.importWalletBackup(payload, passphrase: passphrase, error: &nsError)
      if let error = nsError { throw error }
      return value
    }
  }
}

File: app/modules/pocket-module/android/src/main/java/pocket/module/mindsgn/PocketModule.kt
package pocket.module.mindsgn

import core.WalletCore
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class PocketModule : Module() {
  private val walletCore = WalletCore()

  private val secureKeyStore by lazy {
    val context = requireNotNull(appContext.reactContext) {
      "React context is unavailable"
    }
    SecureKeyStore(context.applicationContext)
  }

  override fun definition() = ModuleDefinition {
    Name("PocketCore")

    AsyncFunction("initWallet") { dataDir: String, password: String, masterKeyB64: String, kdfSaltB64: String ->
      walletCore.init(dataDir, password, masterKeyB64, kdfSaltB64)
    }

    AsyncFunction("initWalletSecure") { dataDir: String, password: String ->
      val masterKeyB64 = secureKeyStore.getOrCreateMasterKeyBase64()
      val kdfSaltB64 = secureKeyStore.getOrCreateKdfSaltBase64()
      walletCore.init(dataDir, password, masterKeyB64, kdfSaltB64)
    }

    AsyncFunction("closeWallet") {
      walletCore.close()
    }

    AsyncFunction("createEthereumWallet") { name: String ->
      walletCore.createEthereumWallet(name)
    }

    AsyncFunction("openOrCreateWallet") { name: String ->
      walletCore.openOrCreateWallet(name)
    }

    AsyncFunction("getBalance") { network: String ->
      walletCore.getBalance(network)
    }

    AsyncFunction("getAccountSummary") { network: String ->
      walletCore.getAccountSummary(network)
    }

    AsyncFunction("getAccountSnapshot") { network: String ->
      walletCore.getAccountSnapshot(network)
    }

    AsyncFunction("createSmartContractAccount") { network: String ->
      walletCore.createSmartContractAccount(network)
    }

    AsyncFunction("getSmartContractAccount") { network: String ->
      walletCore.getSmartContractAccount(network)
    }

    AsyncFunction("listAccounts") {
      walletCore.listAccounts()
    }

    AsyncFunction("sendUsdc") { network: String, destination: String, amount: String, note: String, providerID: String ->
      walletCore.sendUsdc(network, destination, amount, note, providerID)
    }

    AsyncFunction("sendToken") { network: String, tokenIdentifier: String, destination: String, amount: String, note: String, providerID: String ->
      walletCore.sendToken(network, tokenIdentifier, destination, amount, note, providerID)
    }

    AsyncFunction("getUsdcTransactions") { network: String, limit: Int, offset: Int ->
      walletCore.listUsdcTransactions(network, limit.toLong(), offset.toLong())
    }

    AsyncFunction("getTokenTransactions") { network: String, tokenIdentifier: String, limit: Int, offset: Int ->
      walletCore.listTokenTransactions(network, tokenIdentifier, limit.toLong(), offset.toLong())
    }

    AsyncFunction("listAllTransactions") { network: String, limit: Int, offset: Int ->
      walletCore.listAllTransactions(network, limit.toLong(), offset.toLong())
    }

    AsyncFunction("exportBackup") { passphrase: String ->
      walletCore.exportWalletBackup(passphrase)
    }

    AsyncFunction("importBackup") { payload: String, passphrase: String ->
      walletCore.importWalletBackup(payload, passphrase)
    }
  }
}
